### PR TITLE
Bot questing fixes

### PR DESCRIPTION
**Pull request by atidot3**
Remade the PR after the codeline changes that made alot of mess with merge conflict

Bot should be way more consistent and accurate at doing quests.
Add missing implementation of update kill / update loot on quests.
Add "debug quest" strategy to allow debugging.
Bot will always try to keep a certain amount of quests in their log grey / green / red but if they have more than 15 quests they will start to cleanup quests based on priority.
This action will not be performed automatically if the bot has a master or if the bot is an alt.

Known issue:
Gameobject quest arent functionnal yet

image
image
image
image
image
image

### CODE REVIEW

File: src/PlayerbotAI.h [hunk 1]
```diff
@@ -331,16 +383,14 @@ class PlayerbotAI : public PlayerbotAIBase
...
     void HandleCommand(uint32 type, std::string const text, Player* fromPlayer);
-    void QueueChatResponse(uint8 msgtype, ObjectGuid guid1, ObjectGuid guid2, std::string message, std::string chanName,
-                           std::string name);
-    void HandleBotOutgoingPacket(WorldPacket const& packet);
+    void QueueChatResponse(const ChatQueuedReply reply);
+	void HandleBotOutgoingPacket(WorldPacket const& packet);
     void HandleMasterIncomingPacket(WorldPacket const& packet);
     void HandleMasterOutgoingPacket(WorldPacket const& packet);
-    void HandleTeleportAck();
+	void HandleTeleportAck();
```
Comment (atidot3) [inline src/PlayerbotAI.h]: Indentation

File: src/PlayerbotAI.h [hunk 2]
```diff
@@ -486,23 +549,34 @@ class PlayerbotAI : public PlayerbotAIBase
...
     bool IsTellAllowed(PlayerbotSecurityLevel securityLevel = PLAYERBOT_SECURITY_ALLOW_ALL);
 
+    void HandleCommands();
+    void HandleCommand(uint32 type, const std::string& text, Player& fromPlayer, const uint32 lang = LANG_UNIVERSAL);
+
 protected:
-    Player* bot;
-    Player* master;
-    uint32 accountId;
+	Player* bot;
```
Comment (atidot3) [inline src/PlayerbotAI.h]: Indentation

**Comment by noisiver** [comment]
I'm compiling it right now and will be testing it a bit, as usual without them teleporting around.

**Comment by liyunfan1223** [comment]
I started running some tests, and it seems manual SQL import is required, otherwise, the following error occurs. Can the SQL updates be applied automatically? As far as I know, putting the updated SQL file into the update folder should work.

```
Can't get bot text broadcast_quest_accepted_generic! No bots texts for this name!
Can't get bot text broadcast_quest_accepted_generic! No bots texts for this name!
Can't get bot text broadcast_looting_item_normal! No bots texts for this name!
```

**Comment by noisiver** [comment]
I don't know why but they barely do quests on my end anymore. I have a level 19 bot that has 1 quest in their log and 0 rewarded.

I have a grand total of 68 rows in `character_queststatus_rewarded` and 1130 in `character_queststatus`.

Am I missing something from this?
`AiPlayerbot.RandomBotNonCombatStrategies ="+quest,+travel,+rpg,+guild,+debug quest"`

**Comment by noisiver** [comment]
Maybe questing should be a much higher priority of they're below max level? That level 19 bot struggled like crazy to kill a mob because they had no gear because they didn't complete any quests and didn't get lucky with drops.

**Comment by atidot3** [comment]
Yea it will need tweaks for sure, but i still havent found any way to make them do only quests atm

**Comment by noisiver** [comment]
There's no rush, you've done a fantastic job already and I'm sure you'll figure it out in due time.

**Comment by zhangbo8418** [comment]
> Yea it will need tweaks for sure, but i still havent found any way to make them do only quests atm

The robot whispers, and the language is always English.

**Comment by liyunfan1223** [comment]
I'd like to know if more people have tested this PR. It would be great if you could provide some feedback, including confirmation that everything is working properly. Thank you for your efforts.

**Comment by zhangbo8418** [comment]
> I'd like to know if more people have tested this PR. It would be great if you could provide some feedback, including confirmation that everything is working properly. Thank you for your efforts.

The recent update broke the robot language again, the dbc settings are invalid and it keeps speaking English. You can no longer take a boat. There was a previous restoration that allowed you to take boats, subways, and elevators.

**Comment by atidot3** [comment]
Im sorry but you must have done something wrong to your config or database because most of the chat talk they make are based on the DBC settings you are using, there is indeed some hardcoded english chat but its mostly reaction chat.
image


For the boat and elevator its working as expected :
image
image
image
image


are you sure you using a frash core and pr build ?

**Comment by zhangbo8418** [comment]
> Im sorry but you must have done something wrong to your config or database because most of the chat talk they make are based on the DBC settings you are using, there is indeed some hardcoded english chat but its mostly reaction chat. image
> 
> For the boat and elevator its working as expected : image image image image
> 
> are you sure you using a frash core and pr build ?

What I’m talking about is not the pr, but the master.

**Comment by atidot3** [comment]
Can you then please keep this chat related to that pr issue only ?

**Comment by zhangbo8418** [comment]
> Can you then please keep this chat related to that pr issue only ?

Because I also tested this PR, its whispers were always in English.

**Comment by atidot3** [comment]
image
image
image

There is several hardcoded 'helper' just to let the bot master know if the bot has accepted or not a quest as exemple which will always stay like so.

But most of the random talk bots do are using the database:
image
image

You may still need to update the sql files with your own locale translation

File: src/strategy/generic/WorldPacketHandlerStrategy.cpp [hunk 3]
```diff
@@ -34,66 +34,44 @@ void WorldPacketHandlerStrategy::InitTriggers(std::vector<TriggerNode*>& trigger
...
     triggers.push_back(new TriggerNode("xpgain", NextAction::array(0, new NextAction("xp gain", relevance), nullptr)));
-    triggers.push_back(
-        new TriggerNode("levelup", NextAction::array(0, new NextAction("auto teleport for level", relevance + 3),
-                                                     new NextAction("auto talents", relevance + 2),
-                                                     new NextAction("auto learn spell", relevance + 1),
-                                                     new NextAction("auto upgrade equip", relevance), nullptr)));
-    // triggers.push_back(new TriggerNode("group destroyed", NextAction::array(0, new NextAction("reset botAI",
-    // relevance), nullptr)));
-    triggers.push_back(new TriggerNode(
-        "questgiver quest details", NextAction::array(0, new NextAction("turn in query quest", relevance), nullptr)));
```
Comment (liyunfan1223) [inline src/strategy/generic/WorldPacketHandlerStrategy.cpp:78]: The removal will cause bots to be unable to complete quests that end with interacting with a gameobject. An example is Find the Lost Guards https://www.wowhead.com/wotlk/quest=37/find-the-lost-guards

Comment (atidot3) [reply src/strategy/generic/WorldPacketHandlerStrategy.cpp:78]: Will uncomment it wasnt sure it was used.

File: src/strategy/actions/WorldPacketActionContext.h [hunk 4]
```diff
@@ -93,8 +102,6 @@ class WorldPacketActionContext : public NamedObjectContext<Action>
         creators["lfg teleport"] = &WorldPacketActionContext::lfg_teleport;
         creators["see spell"] = &WorldPacketActionContext::see_spell;
         creators["arena team accept"] = &WorldPacketActionContext::arena_team_accept;
-        creators["turn in query quest"] = &WorldPacketActionContext::turn_in_query_quest;
```
Comment (liyunfan1223) [inline src/strategy/actions/WorldPacketActionContext.h:96]: Also revert here :D

Comment (atidot3) [reply src/strategy/actions/WorldPacketActionContext.h:96]: Ooops done

**Comment by liyunfan1223** [comment]
I believe it’s okay to merge, but to be honest, I can’t guarantee there won’t be issues. What I strongly recommend is to split your PR into as many small parts as possible and ensure each part do **_only one_** specific task and clearly lists all consequences. Handling such a large PR is exhausting, and I’d prefer not to go through this again.

**Comment by liyunfan1223** [comment]
I noticed that the server diff increased significantly after this PR was merged. The reason is the modification you made to src/strategy/actions/ChooseRpgTargetAction.cpp.

I post the performance testing results. The reverted version (perf_revert.svg) only reverted ChooseRpgTargetAction.cpp, and the performance difference should be noticeable (the propotion of ChooseRpgTargetAction increased from 7.5% to 12.63%). I ran 800 bots in debug mode and the average diff increased from 35ms to 60ms after running 10 mins (for a stable diff).

I'm curious about why you made this modification on ChooseRpgTargetAction.cpp and can it be reverted?

perf_pr433.svg
perf_pr_10min


perf_revert.svg
perf_reverted

**Pull request merged**